<!DOCTYPE html>





<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32.ico?v=7.3.0">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16.ico?v=7.3.0">

<link rel="stylesheet" href="/css/main.css?v=7.3.0">


<link rel="stylesheet" href="/lib/font-awesome/css/font-awesome.min.css?v=4.7.0">


<script id="hexo-configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Mist',
    version: '7.3.0',
    exturl: false,
    sidebar: {"position":"left","display":"post","offset":12,"onmobile":false},
    copycode: {"enable":false,"show_result":false,"style":null},
    back2top: {"enable":true,"sidebar":false,"scrollpercent":false},
    bookmark: {"enable":false,"color":"#222","save":"auto"},
    fancybox: false,
    mediumzoom: false,
    lazyload: false,
    pangu: false,
    algolia: {
      appID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    },
    localsearch: {"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},
    path: 'search.xml',
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    translation: {
      copy_button: '复制',
      copy_success: '复制成功',
      copy_failure: '复制失败'
    },
    sidebarPadding: 40
  };
</script>

  <meta name="description" content="如何提取拟南芥（或非模式物种）基因的启动子序列_基于R">
<meta name="keywords" content="Bioinformatics,R,Sequence_operation,Range_operation">
<meta property="og:type" content="article">
<meta property="og:title" content="如何提取基因的启动子序列_基于R">
<meta property="og:url" content="http://yoursite.com/2019/12/29/如何提取基因的启动子序列-基于R/index.html">
<meta property="og:site_name" content="Shawn&#39;s Blog">
<meta property="og:description" content="如何提取拟南芥（或非模式物种）基因的启动子序列_基于R">
<meta property="og:locale" content="zh-CN">
<meta property="og:updated_time" content="2019-12-29T07:34:22.107Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="如何提取基因的启动子序列_基于R">
<meta name="twitter:description" content="如何提取拟南芥（或非模式物种）基因的启动子序列_基于R">
  <link rel="canonical" href="http://yoursite.com/2019/12/29/如何提取基因的启动子序列-基于R/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome: false,
    isPost: true,
    isPage: false,
    isArchive: false
  };
</script>

  <title>如何提取基因的启动子序列_基于R | Shawn's Blog</title>
  <meta name="generator" content="Hexo 3.9.0">
  








  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .logo,
  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-CN">
  <div class="container use-motion">
    <div class="headband"></div>
	<a href="https://github.com/shangguandong1996" class="github-corner" aria-label="View source on GitHub"><svg width="80" height="80" viewBox="0 0 250 250" style="fill:#151513; color:#fff; position: absolute; top: 0; border: 0; right: 0;" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path></svg></a><style>.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}</style>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-meta">

    <div>
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Shawn's Blog</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
  </div>

  <div class="site-nav-toggle">
    <button aria-label="切换导航栏">
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>


<nav class="site-nav">
  
  <ul id="menu" class="menu">
      
      
      
        
        <li class="menu-item menu-item-home">
      
    

    <a href="/" rel="section"><i class="menu-item-icon fa fa-fw fa-home"></i> <br>首页</a>

  </li>
      
      
      
        
        <li class="menu-item menu-item-about">
      
    

    <a href="/about/" rel="section"><i class="menu-item-icon fa fa-fw fa-user"></i> <br>关于</a>

  </li>
      
      
      
        
        <li class="menu-item menu-item-tags">
      
    

    <a href="/tags/" rel="section"><i class="menu-item-icon fa fa-fw fa-tags"></i> <br>标签</a>

  </li>
      
      
      
        
        <li class="menu-item menu-item-categories">
      
    

    <a href="/categories/" rel="section"><i class="menu-item-icon fa fa-fw fa-th"></i> <br>分类</a>

  </li>
      
      
      
        
        <li class="menu-item menu-item-archives">
      
    

    <a href="/archives/" rel="section"><i class="menu-item-icon fa fa-fw fa-archive"></i> <br>归档</a>

  </li>
      <li class="menu-item menu-item-search">
        <a href="javascript:;" class="popup-trigger">
        
          <i class="menu-item-icon fa fa-search fa-fw"></i> <br>搜索</a>
      </li>
    
  </ul>

    

</nav>
  <div class="site-search">
    <div class="popup search-popup">
    <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocorrect="off" autocapitalize="none"
           placeholder="搜索..." spellcheck="false"
           type="text" id="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result"></div>

</div>
<div class="search-pop-overlay"></div>

  </div>
</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
      <article itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block post">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/12/29/如何提取基因的启动子序列-基于R/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Just For Daisy">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Shawn's Blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">如何提取基因的启动子序列_基于R

          
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>

              
                
              

              <time title="创建时间：2019-12-29 15:31:30 / 修改时间：15:34:22" itemprop="dateCreated datePublished" datetime="2019-12-29T15:31:30+08:00">2019-12-29</time>
            </span>
          
            

            
          
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              <span class="post-meta-item-text">分类于</span>
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing"><a href="/categories/生信操作/" itemprop="url" rel="index"><span itemprop="name">生信操作</span></a></span>

                
                
              
            </span>
          

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <p>如何提取拟南芥（或非模式物种）基因的启动子序列_基于R</p>
<a id="more"></a>
<p>对于提取序列来说，我们一般会选择用samtools、bedtools等，因为他们有非常完善的操作<strong>序列</strong>的功能，比如取两个bed文件的交集，提取某段序列等等。但实际上，他们对于<strong>区间</strong>的操作是不如R方便的，比如取基因body，取第一个外显子，取启动子区域等等。所以，我们有时候就会想能否在R里面，花式找了些区间之后，提取他们的序列，实际上，这也是可以的，需要分成几步。</p>
<p>首先我们需要先加载一些包</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># BSgenome.Athaliana.TAIR.TAIR9是拟南芥的基因组序列包</span></span><br><span class="line"><span class="comment"># TAIR9和TAIR10是一样的，无非是TAIR10自身的基因注释丰富了些</span></span><br><span class="line"><span class="keyword">library</span>(BSgenome.Athaliana.TAIR.TAIR9)</span><br><span class="line"></span><br><span class="line"><span class="comment"># TxDb.Athaliana.BioMart.plantsmart28是拟南芥的Txdb包</span></span><br><span class="line"><span class="comment"># 就是我们ChIPSeeker做peak注释的那个Txdb</span></span><br><span class="line"><span class="keyword">library</span>(TxDb.Athaliana.BioMart.plantsmart28)</span><br><span class="line">Txdb_package &lt;- TxDb.Athaliana.BioMart.plantsmart28</span><br><span class="line"></span><br><span class="line"><span class="comment"># 也可以用自己的gtf做</span></span><br><span class="line"><span class="comment"># 所以这些操作也可以针对非模式物种</span></span><br><span class="line">Txdb_gtf &lt;- makeTxDbFromGFF(<span class="string">"~/reference/annoation/Athaliana/Araport11/Araport11_GFF3_genes_transposons.201606.gtf"</span>)</span><br></pre></td></tr></table></figure>
<p>然后是提取基因的区间</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br></pre></td><td class="code"><pre><span class="line">&gt; gene_package &lt;- genes(Txdb_package)</span><br><span class="line">&gt; gene_package</span><br><span class="line">GRanges object with <span class="number">33602</span> ranges and <span class="number">1</span> metadata column:</span><br><span class="line">            seqnames        ranges strand |     gene_id</span><br><span class="line">               &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span><br><span class="line">  AT1G01010        <span class="number">1</span>     <span class="number">3631</span>-<span class="number">5899</span>      + |   AT1G01010</span><br><span class="line">  AT1G01020        <span class="number">1</span>     <span class="number">5928</span>-<span class="number">8737</span>      - |   AT1G01020</span><br><span class="line">  AT1G01030        <span class="number">1</span>   <span class="number">11649</span>-<span class="number">13714</span>      - |   AT1G01030</span><br><span class="line">  AT1G01040        <span class="number">1</span>   <span class="number">23146</span>-<span class="number">31227</span>      + |   AT1G01040</span><br><span class="line">  AT1G01046        <span class="number">1</span>   <span class="number">28500</span>-<span class="number">28706</span>      + |   AT1G01046</span><br><span class="line">        <span class="keyword">...</span>      <span class="keyword">...</span>           <span class="keyword">...</span>    <span class="keyword">...</span> .         <span class="keyword">...</span></span><br><span class="line">  ATMG01370       Mt <span class="number">360717</span>-<span class="number">361052</span>      - |   ATMG01370</span><br><span class="line">  ATMG01380       Mt <span class="number">361062</span>-<span class="number">361179</span>      - |   ATMG01380</span><br><span class="line">  ATMG01390       Mt <span class="number">361350</span>-<span class="number">363284</span>      - |   ATMG01390</span><br><span class="line">  ATMG01400       Mt <span class="number">363725</span>-<span class="number">364042</span>      + |   ATMG01400</span><br><span class="line">  ATMG01410       Mt <span class="number">366086</span>-<span class="number">366700</span>      - |   ATMG01410</span><br><span class="line">  -------</span><br><span class="line">  seqinfo: <span class="number">7</span> sequences (<span class="number">1</span> circular) from an unspecified genome</span><br><span class="line"></span><br><span class="line"><span class="comment"># 注意这里有个no seqlengths</span></span><br><span class="line">&gt; gene_gtf &lt;- genes(Txdb_gtf)</span><br><span class="line">&gt; gene_gtf</span><br><span class="line">GRanges object with <span class="number">37330</span> ranges and <span class="number">1</span> metadata column:</span><br><span class="line">            seqnames        ranges strand |     gene_id</span><br><span class="line">               &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span><br><span class="line">  AT1G01010     Chr1     <span class="number">3631</span>-<span class="number">5899</span>      + |   AT1G01010</span><br><span class="line">  AT1G01020     Chr1     <span class="number">6788</span>-<span class="number">9130</span>      - |   AT1G01020</span><br><span class="line">  AT1G01030     Chr1   <span class="number">11649</span>-<span class="number">13714</span>      - |   AT1G01030</span><br><span class="line">  AT1G01040     Chr1   <span class="number">23121</span>-<span class="number">31227</span>      + |   AT1G01040</span><br><span class="line">  AT1G01050     Chr1   <span class="number">31170</span>-<span class="number">33171</span>      - |   AT1G01050</span><br><span class="line">        <span class="keyword">...</span>      <span class="keyword">...</span>           <span class="keyword">...</span>    <span class="keyword">...</span> .         <span class="keyword">...</span></span><br><span class="line">  ATMG09730     ChrM <span class="number">181268</span>-<span class="number">181347</span>      + |   ATMG09730</span><br><span class="line">  ATMG09740     ChrM <span class="number">191954</span>-<span class="number">192025</span>      - |   ATMG09740</span><br><span class="line">  ATMG09950     ChrM <span class="number">333651</span>-<span class="number">333725</span>      - |   ATMG09950</span><br><span class="line">  ATMG09960     ChrM <span class="number">347266</span>-<span class="number">347348</span>      + |   ATMG09960</span><br><span class="line">  ATMG09980     ChrM <span class="number">359666</span>-<span class="number">359739</span>      - |   ATMG09980</span><br><span class="line">  -------</span><br><span class="line">  seqinfo: <span class="number">7</span> sequences (<span class="number">2</span> circular) from an unspecified genome; no seqlengths</span><br></pre></td></tr></table></figure>
<p>然后根据基因区域来提取启动子序列</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">&gt; promoter_package &lt;- promoters(gene_package,upstream = <span class="number">4000</span>,downstream = <span class="number">500</span>)</span><br><span class="line">Warning message:</span><br><span class="line">In valid.GenomicRanges.seqinfo(x, suggest.trim = <span class="literal">TRUE</span>) :</span><br><span class="line">  GRanges object contains <span class="number">11</span> out-of-bound ranges located on sequences <span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>, <span class="number">4</span>, and Pt. Note that</span><br><span class="line">  ranges located on a sequence whose length is unknown (<span class="literal">NA</span>) or on a circular sequence are not</span><br><span class="line">  considered out-of-bound (use seqlengths() and isCircular() to get the lengths and circularity flags</span><br><span class="line">  of the underlying sequences). You can use trim() to trim these ranges. See</span><br><span class="line">  ?`trim,GenomicRanges-method` <span class="keyword">for</span> more information.</span><br><span class="line"></span><br><span class="line">&gt; promoter_gtf &lt;- promoters(gene_gtf,upstream = <span class="number">4000</span>,downstream = <span class="number">500</span>)</span><br></pre></td></tr></table></figure>
<p>这里就出现了一个坑了，就是你在用Txdb包的时候，出现了warning，但用自建gtf构建的Txdb没有出现了warning。实际上，这种warning是因为你的区间取过界了。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">&gt; promoter_package</span><br><span class="line">GRanges object with <span class="number">33602</span> ranges and <span class="number">1</span> metadata column:</span><br><span class="line">            seqnames        ranges strand |     gene_id</span><br><span class="line">               &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span><br><span class="line">  AT1G01010        <span class="number">1</span>     -<span class="number">369</span>-<span class="number">4130</span>      + |   AT1G01010</span><br><span class="line">  AT1G01020        <span class="number">1</span>    <span class="number">8238</span>-<span class="number">12737</span>      - |   AT1G01020</span><br><span class="line">  AT1G01030        <span class="number">1</span>   <span class="number">13215</span>-<span class="number">17714</span>      - |   AT1G01030</span><br><span class="line">  AT1G01040        <span class="number">1</span>   <span class="number">19146</span>-<span class="number">23645</span>      + |   AT1G01040</span><br><span class="line">  AT1G01046        <span class="number">1</span>   <span class="number">24500</span>-<span class="number">28999</span>      + |   AT1G01046</span><br><span class="line">        <span class="keyword">...</span>      <span class="keyword">...</span>           <span class="keyword">...</span>    <span class="keyword">...</span> .         <span class="keyword">...</span></span><br><span class="line">  ATMG01370       Mt <span class="number">360553</span>-<span class="number">365052</span>      - |   ATMG01370</span><br><span class="line">  ATMG01380       Mt <span class="number">360680</span>-<span class="number">365179</span>      - |   ATMG01380</span><br><span class="line">  ATMG01390       Mt <span class="number">362785</span>-<span class="number">367284</span>      - |   ATMG01390</span><br><span class="line">  ATMG01400       Mt <span class="number">359725</span>-<span class="number">364224</span>      + |   ATMG01400</span><br><span class="line">  ATMG01410       Mt <span class="number">366201</span>-<span class="number">370700</span>      - |   ATMG01410</span><br><span class="line">  -------</span><br><span class="line">  seqinfo: <span class="number">7</span> sequences (<span class="number">1</span> circular) from an unspecified genome</span><br></pre></td></tr></table></figure>
<p>而promoter_gtf没有报错是因为他不知道<strong>你的染色体的长度范围在哪里</strong>。</p>
<p>所以下面我的操作都是基于promoter_gtf，这样还可以应用在<strong>非模式物种</strong>里面。首先你在构建Txdb的时候就需要给他传进去染色体长度了。但你怎么知道每条染色体是多长呢，这就需要你的fai文件了。fai文件是给基因组fa做索引的时候出现的，没有的话你可以直接用samtools faidx做下。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">$ samtools faidx Athaliana.fa</span><br><span class="line"></span><br><span class="line"><span class="comment"># 这里的第2列是染色体长度</span></span><br><span class="line">Chr1    30427671        6       79      80</span><br><span class="line">Chr2    19698289        30812844        79      80</span><br><span class="line">Chr3    23459830        50760485        79      80</span><br><span class="line">Chr4    18585056        74517281        79      80</span><br><span class="line">Chr5    26975502        93337597        79      80</span><br><span class="line">ChrM    366924  120654568       79      80</span><br><span class="line">ChrC    154478  121026144       79      80</span><br></pre></td></tr></table></figure>
<p>然后我们就可以在构建的时候手动输进对应的长度了，但这只是拟南芥这个组装完整的基因组，如果是一些非模式物种，scaffold级别的话，就会有数百条染色体长度了，根本不可能手动输。所以我们还需要读进R里面进行一些操作。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 注意不要变成因子</span></span><br><span class="line">&gt; seq_length &lt;- read.table(<span class="string">"~/reference/genome/TAIR10/Athaliana.fa.fai"</span>,</span><br><span class="line">+                          stringsAsFactors = <span class="literal">F</span>)[,<span class="number">1</span>:<span class="number">2</span>]</span><br><span class="line">&gt; seq_length$V1</span><br><span class="line">[<span class="number">1</span>] <span class="string">"Chr1"</span> <span class="string">"Chr2"</span> <span class="string">"Chr3"</span> <span class="string">"Chr4"</span> <span class="string">"Chr5"</span> <span class="string">"ChrM"</span> <span class="string">"ChrC"</span></span><br><span class="line">&gt; seq_length$V2</span><br><span class="line">[<span class="number">1</span>] <span class="number">30427671</span> <span class="number">19698289</span> <span class="number">23459830</span> <span class="number">18585056</span> <span class="number">26975502</span>   <span class="number">366924</span>   <span class="number">154478</span></span><br><span class="line">&gt; Txdb_gtf &lt;- makeTxDbFromGFF(<span class="string">"~/reference/annoation/Athaliana/Araport11/Araport11_GFF3_genes_transposons.201606.gtf"</span>,</span><br><span class="line">+                             chrominfo = Seqinfo(seqnames = seq_length$V1,</span><br><span class="line">+                                                 seqlengths = seq_length$V2))</span><br></pre></td></tr></table></figure>
<p>这样我们提取启动子序列的时候就会报错了</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br></pre></td><td class="code"><pre><span class="line">&gt; gene_gtf &lt;- genes(Txdb_gtf)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 此时no seqlengths消失了</span></span><br><span class="line">&gt; gene_gtf</span><br><span class="line">GRanges object with <span class="number">37330</span> ranges and <span class="number">1</span> metadata column:</span><br><span class="line">            seqnames        ranges strand |     gene_id</span><br><span class="line">               &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span><br><span class="line">  AT1G01010     Chr1     <span class="number">3631</span>-<span class="number">5899</span>      + |   AT1G01010</span><br><span class="line">  AT1G01020     Chr1     <span class="number">6788</span>-<span class="number">9130</span>      - |   AT1G01020</span><br><span class="line">  AT1G01030     Chr1   <span class="number">11649</span>-<span class="number">13714</span>      - |   AT1G01030</span><br><span class="line">  AT1G01040     Chr1   <span class="number">23121</span>-<span class="number">31227</span>      + |   AT1G01040</span><br><span class="line">  AT1G01050     Chr1   <span class="number">31170</span>-<span class="number">33171</span>      - |   AT1G01050</span><br><span class="line">        <span class="keyword">...</span>      <span class="keyword">...</span>           <span class="keyword">...</span>    <span class="keyword">...</span> .         <span class="keyword">...</span></span><br><span class="line">  ATMG09730     ChrM <span class="number">181268</span>-<span class="number">181347</span>      + |   ATMG09730</span><br><span class="line">  ATMG09740     ChrM <span class="number">191954</span>-<span class="number">192025</span>      - |   ATMG09740</span><br><span class="line">  ATMG09950     ChrM <span class="number">333651</span>-<span class="number">333725</span>      - |   ATMG09950</span><br><span class="line">  ATMG09960     ChrM <span class="number">347266</span>-<span class="number">347348</span>      + |   ATMG09960</span><br><span class="line">  ATMG09980     ChrM <span class="number">359666</span>-<span class="number">359739</span>      - |   ATMG09980</span><br><span class="line">  -------</span><br><span class="line">  seqinfo: <span class="number">7</span> sequences from an unspecified genome</span><br><span class="line"></span><br><span class="line"><span class="comment"># 出现了报错</span></span><br><span class="line">&gt; promoter_gtf &lt;- promoters(gene_gtf,upstream = <span class="number">4000</span>,downstream = <span class="number">500</span>)</span><br><span class="line">Warning message:</span><br><span class="line">In valid.GenomicRanges.seqinfo(x, suggest.trim = <span class="literal">TRUE</span>) :</span><br><span class="line">  GRanges object contains <span class="number">15</span> out-of-bound ranges located on sequences Chr1, Chr2, Chr3, Chr4, Chr5, ChrC, and ChrM.</span><br><span class="line">  Note that ranges located on a sequence whose length is unknown (<span class="literal">NA</span>) or on a circular sequence are not considered</span><br><span class="line">  out-of-bound (use seqlengths() and isCircular() to get the lengths and circularity flags of the underlying</span><br><span class="line">  sequences). You can use trim() to trim these ranges. See ?`trim,GenomicRanges-method` <span class="keyword">for</span> more information.</span><br><span class="line"></span><br><span class="line"><span class="comment"># 我们用trim函数去掉越过界的区间</span></span><br><span class="line"><span class="comment"># 原本过界的就变成了1了</span></span><br><span class="line">&gt; promoter_gtf &lt;- trim(promoter_gtf)</span><br><span class="line">&gt; promoter_gtf</span><br><span class="line">GRanges object with <span class="number">37330</span> ranges and <span class="number">1</span> metadata column:</span><br><span class="line">            seqnames        ranges strand |     gene_id</span><br><span class="line">               &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span><br><span class="line">  AT1G01010     Chr1        <span class="number">1</span>-<span class="number">4130</span>      + |   AT1G01010</span><br><span class="line">  AT1G01020     Chr1    <span class="number">8631</span>-<span class="number">13130</span>      - |   AT1G01020</span><br><span class="line">  AT1G01030     Chr1   <span class="number">13215</span>-<span class="number">17714</span>      - |   AT1G01030</span><br><span class="line">  AT1G01040     Chr1   <span class="number">19121</span>-<span class="number">23620</span>      + |   AT1G01040</span><br><span class="line">  AT1G01050     Chr1   <span class="number">32672</span>-<span class="number">37171</span>      - |   AT1G01050</span><br><span class="line">        <span class="keyword">...</span>      <span class="keyword">...</span>           <span class="keyword">...</span>    <span class="keyword">...</span> .         <span class="keyword">...</span></span><br><span class="line">  ATMG09730     ChrM <span class="number">177268</span>-<span class="number">181767</span>      + |   ATMG09730</span><br><span class="line">  ATMG09740     ChrM <span class="number">191526</span>-<span class="number">196025</span>      - |   ATMG09740</span><br><span class="line">  ATMG09950     ChrM <span class="number">333226</span>-<span class="number">337725</span>      - |   ATMG09950</span><br><span class="line">  ATMG09960     ChrM <span class="number">343266</span>-<span class="number">347765</span>      + |   ATMG09960</span><br><span class="line">  ATMG09980     ChrM <span class="number">359240</span>-<span class="number">363739</span>      - |   ATMG09980</span><br><span class="line">  -------</span><br><span class="line">  seqinfo: <span class="number">7</span> sequences from an unspecified genome</span><br></pre></td></tr></table></figure>
<p>然后我们只取前2个基因出来看看效果</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 这边的GRange是带名字的GRange，所以我们取区间的时候可以用基因名字</span></span><br><span class="line">&gt; promoter_gtf_part &lt;- promoter_gtf[c(<span class="string">"AT1G01010"</span>,<span class="string">"AT1G01020"</span>)]</span><br><span class="line">&gt; promoter_gtf_part</span><br><span class="line">GRanges object with <span class="number">2</span> ranges and <span class="number">1</span> metadata column:</span><br><span class="line">            seqnames     ranges strand |     gene_id</span><br><span class="line">               &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span><br><span class="line">  AT1G01010     Chr1     <span class="number">1</span>-<span class="number">4130</span>      + |   AT1G01010</span><br><span class="line">  AT1G01020     Chr1 <span class="number">8631</span>-<span class="number">13130</span>      - |   AT1G01020</span><br><span class="line">  -------</span><br><span class="line">  seqinfo: <span class="number">7</span> sequences from an unspecified genome</span><br><span class="line"></span><br><span class="line"><span class="comment"># 用的是Biostrings包的getSeq</span></span><br><span class="line"><span class="comment"># 注意你这里的Grange的seqnames要和genome的seqnames要一致</span></span><br><span class="line"><span class="comment"># 有时候Grange是1，而Genome是Chr1</span></span><br><span class="line">&gt; getSeq(BSgenome.Athaliana.TAIR.TAIR9,promoter_gtf_part)</span><br><span class="line">  A DNAStringSet instance of length <span class="number">2</span></span><br><span class="line">    width seq                                                                                               names               </span><br><span class="line">[<span class="number">1</span>]  <span class="number">4130</span> CCCTAAACCCTAAACCCTAAACCCTAAACCTCTGAATCCTTAATCCC...GGTTTCTGGTAAATGGAAGCTTACCGGAGAATCTGTTGAGGTCAAGG AT1G01010</span><br><span class="line">[<span class="number">2</span>]  <span class="number">4500</span> ATCCGCAACAATTCACCAATTGAAGAACAAGAGAAAGGTTTAAACTT...GAGAGAGAGCAATGGCGGCGAGTGAACACAGATGCGTGGGATGTGGT AT1G01020</span><br><span class="line"></span><br><span class="line">promoter_gtf_part_seq &lt;- getSeq(BSgenome.Athaliana.TAIR.TAIR9,promoter_gtf_part)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 然后用writeXStringSet写成fa</span></span><br><span class="line">writeXStringSet(promoter_gtf_part_seq,</span><br><span class="line">                filepath = <span class="string">"test.fasta"</span>,</span><br><span class="line">                format = <span class="string">"fasta"</span>)</span><br></pre></td></tr></table></figure>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">&gt;AT1G01010</span><br><span class="line">CCCTAAACCCTAAACCCTAAACCCTAAACCTCTGAATCCTTAATCCCTAAATCCCTAAATCTTTAAATCCTACATCCATG</span><br><span class="line">AATCCCTAAATACCTAATTCCCTAAACCCGAAACCGGTTTCTCTGGTTGAAAATCATTGTGTATATAATGATAATTTTAT</span><br><span class="line">CGTTTTTATGTAATTGCTTATTGTTGTGTGTAGATTTTTTAAAAATATCATTTGAGGTCAATACAAATCCTATTTCTTGT</span><br><span class="line">GGTTTTCTTTCCTTCACTTAGCTATGGATGGTTTATCTTCATTTGTTATATTGGATACAAGCTTTGCTACGATCTACATT</span><br><span class="line">TGGGAATGTGAGTCTCTTATTGTAACCTTAGGGTTGGTTTATCTCAAGAATCTTATTAATTGTTTGGACTGTTTATGTTT</span><br><span class="line">GGACATTTATTGTCATTCTTACTCCTTTGTGGAAATGTTTGTTCTATCAATTTATCTTTTGTGGGAAAATTATTTAGTTG</span><br></pre></td></tr></table></figure>

    </div>

    
    
    
        
      

      <footer class="post-footer">
          
            
          
          <div class="post-tags">
            
              <a href="/tags/Bioinformatics/" rel="tag"># Bioinformatics</a>
            
              <a href="/tags/R/" rel="tag"># R</a>
            
              <a href="/tags/Sequence-operation/" rel="tag"># Sequence_operation</a>
            
              <a href="/tags/Range-operation/" rel="tag"># Range_operation</a>
            
          </div>
        

        

          <div class="post-nav">
            <div class="post-nav-next post-nav-item">
              
                <a href="/2019/12/27/从Uniprot那里得到对应的拟南芥基因ID/" rel="next" title="从Uniprot那里得到对应的拟南芥基因ID">
                  <i class="fa fa-chevron-left"></i> 从Uniprot那里得到对应的拟南芥基因ID
                </a>
              
            </div>

            <span class="post-nav-divider"></span>

            <div class="post-nav-prev post-nav-item">
              
                <a href="/2019/12/30/R-studio的注释/" rel="prev" title="R studio的注释">
                  R studio的注释 <i class="fa fa-chevron-right"></i>
                </a>
              
            </div>
          </div>
        
      </footer>
    
  </div>
  
  
  
  </article>

  </div>


          </div>
          

        </div>
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">
        
        
        
        
      

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概览
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
  <p class="site-author-name" itemprop="name">Just For Daisy</p>
  <div class="site-description" itemprop="description"></div>
</div>
  <nav class="site-state motion-element">
      <div class="site-state-item site-state-posts">
        
          <a href="/archives/">
        
          <span class="site-state-item-count">34</span>
          <span class="site-state-item-name">日志</span>
        </a>
      </div>
    
      
      
      <div class="site-state-item site-state-categories">
        
          
            <a href="/categories/">
          
        
        
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
        <span class="site-state-item-count">9</span>
        <span class="site-state-item-name">分类</span>
        </a>
      </div>
    
      
      
      <div class="site-state-item site-state-tags">
        
          
            <a href="/tags/">
          
        
        
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
          
        
        <span class="site-state-item-count">10</span>
        <span class="site-state-item-name">标签</span>
        </a>
      </div>
    
  </nav>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2020</span>
  <span class="with-love" id="animate">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Just For Daisy</span>
</div>
  <div class="powered-by">由 <a href="https://hexo.io" class="theme-link" rel="noopener" target="_blank">Hexo</a> 强力驱动 v3.9.0</div>
  <span class="post-meta-divider">|</span>
  <div class="theme-info">主题 – <a href="https://theme-next.org" class="theme-link" rel="noopener" target="_blank">NexT.Mist</a> v7.3.0</div>

        












        
      </div>
    </footer>
  </div>

  


  <script src="/lib/anime.min.js?v=3.1.0"></script>
  <script src="/lib/velocity/velocity.min.js?v=1.2.1"></script>
  <script src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
<script src="/js/utils.js?v=7.3.0"></script><script src="/js/motion.js?v=7.3.0"></script>
<script src="/js/schemes/muse.js?v=7.3.0"></script>

<script src="/js/next-boot.js?v=7.3.0"></script>



  








  <script src="/js/local-search.js?v=7.3.0"></script>














  

  
    
      
<script type="text/x-mathjax-config">

  MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [ ['$', '$'], ['\\(', '\\)'] ],
      processEscapes: true,
      skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
    },
    TeX: {
      equationNumbers: {
        autoNumber: 'AMS'
      }
    }
  });

  MathJax.Hub.Register.StartupHook('TeX Jax Ready', function() {
    MathJax.InputJax.TeX.prefilterHooks.Add(function(data) {
      if (data.display) {
        var next = data.script.nextSibling;
        while (next && next.nodeName.toLowerCase() === '#text') {
          next = next.nextSibling;
        }
        if (next && next.nodeName.toLowerCase() === 'br') {
          next.parentNode.removeChild(next);
        }
      }
    });
  });

  MathJax.Hub.Queue(function() {
    var all = MathJax.Hub.getAllJax(), i;
    for (i = 0; i < all.length; i += 1) {
      element = document.getElementById(all[i].inputID + '-Frame').parentNode;
      if (element.nodeName.toLowerCase() == 'li') {
        element = element.parentNode;
      }
      element.classList.add('has-jax');
    }
  });
</script>
<script>
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/mathjax@2/MathJax.js?config=TeX-AMS-MML_HTMLorMML', () => {
    MathJax.Hub.Typeset();
  }, window.MathJax);
</script>

    
  

  

  

</body>
</html>
